// boomutil/src/elements/dialog/index.js
import { els } from "../../util/general";

export class BoomDialog extends HTMLElement {
	constructor(options) {
		super();
		this.options = options;
		this.innerHTML = this[options.style ?? "message"](options);
	}
	
	message({ title, description, buttons }) {
		return `<div class="message">
			<h1>${title}</h1>
			<p>${description}</p>
			${this.getButtons(buttons)}
		</div>`;
	}
	
	getButtons(buttons) {
		const holder = document.createElement("div");
		holder.className = "buttons";
		for(const { title } of buttons) {
			const button = document.createElement("boom-button");
			button.className = "neon";
			button.textContent = title;
			holder.appendChild(button);
		}
		return holder.outerHTML;
	}
	
	show(parent) {
		(parent || document.body).appendChild(this);
		setTimeout(() => {
			this.classList.add("shown");
			els(this, "boom-button").forEach(item => {
				item.onclick = () => {
					this.close();
				};
			});
		}, 250);
	}
	
	close() {
		this.classList.add("closed");
		setTimeout(() => {
			this.remove();
		}, 500);
	}
}
