// boomutil/src/elemenets/select/select.js
import { els } from "../../util/general";

export default class BoomSelect extends HTMLElement {
	constructor() {
		super();
		this.handler = () => {}
	}
	
	connectedCallback() {
		this.mode = this.getAttribute("mode") || "single";
	}
	
	attributeChangedCallback(key, oldValue, newValue) {
		if(key == "mode") {
			this.mode = newValue;
		}
	}
	
	load(array) {
		this.innerHTML = ((html = "") => {
			for(const tag of array) {
				html += `<boom-item title="${tag.title}" `;
				if("gicon" in tag) html += `gicon="${tag.gicon}"`;
				html += `></boom-item>`;
			}
			return html;
		})();
		this.updateClickListeners();
	}
	
	select(id, item) {
		this.handler(id, item);
		if(this.mode == "single") {
			this.reset();
			if(item) {
				item.classList.add("active");
			} else {
				els(this, "boom-item")[id].classList.add("active");
			}
		}
	}
	
	updateClickListeners() {
		els(this, "boom-item").forEach((item, id) => {
			item.className = "";
			if(this.mode == "single" && id == 0) {
				this.reset();
				item.classList.add("active");
			}
			item.onclick = () => this.select(id, item);
		});
	}
	
	reset() {
		for(const item of els(this, "boom-item")) {
			item.className = "";
		}
	}
	
	set mode(mode) {
		this.modeValue = mode;
		this.updateClickListeners();
	}
	
	get mode() {
		return this.modeValue;
	}
	
	set onSelected(fun) {
		this.handler = fun;
	}
}
